extension/src/service/klarna-service.js

```diff
@@ -0,0 +1,86 @@
+const {
+  ADYEN_PERCENTAGE_MINOR_UNIT,
+  DEFAULT_PAYMENT_LANGUAGE,
+  KLARNA_DEFAULT_LINE_ITEM_NAME
+} = require('../config/klarna-constants')
+
+function createLineItems (payment, cart) {
+  const lineItems = []
+  const locales = _getLocales(cart, payment)
+
+  cart.lineItems.forEach((item) => {
+    lineItems.push(_createAdyenLineItemFromLineItem(item, locales))
+  })
+
+  cart.customLineItems.forEach((item) => {
+    lineItems.push(_createAdyenLineItemFromCustomLineItem(item, locales))
+  })
+
+  lineItems.push(_createShippingInfoAdyenLineItem(cart.shippingInfo))
+
+  return lineItems
+}
+
+function _getLocales (cart, payment) {
+  const locales = []
+  let paymentLanguage = payment.custom && payment.custom.fields['languageCode']
+  if (!paymentLanguage)
+    paymentLanguage = cart.locale
+  if (paymentLanguage)
+    locales.push(paymentLanguage)
+  if (!paymentLanguage || paymentLanguage !== DEFAULT_PAYMENT_LANGUAGE)
+    locales.push(DEFAULT_PAYMENT_LANGUAGE)
+
+  return locales
+}
+
+function _createAdyenLineItemFromLineItem (ctpLineItem, locales) {
+  return {
+    id: ctpLineItem.variant.sku,
+    quantity: ctpLineItem.quantity,
+    description: _localizeOrFallback(ctpLineItem.name, locales, KLARNA_DEFAULT_LINE_ITEM_NAME),
+    amountIncludingTax: ctpLineItem.price.value.centAmount,
+    taxPercentage: ctpLineItem.taxRate.amount * ADYEN_PERCENTAGE_MINOR_UNIT
+  }
+}
+
+function _createAdyenLineItemFromCustomLineItem (ctpLineItem, locales) {
+  return {
+    id: ctpLineItem.id,
+    quantity: ctpLineItem.quantity,
+    description: _localizeOrFallback(ctpLineItem.name, locales, KLARNA_DEFAULT_LINE_ITEM_NAME),
+    amountIncludingTax: ctpLineItem.money.centAmount,
+    taxPercentage: ctpLineItem.taxRate.amount * ADYEN_PERCENTAGE_MINOR_UNIT
+  }
+}
+
+function _createShippingInfoAdyenLineItem (shippingInfo) {
```

Comment:
shipping info is an optional field in the cart,  do we need to do a check here to avoid NPEs?
OK... I could find different places to find checks, here what's the strategy for payment handler, if the cart does not have enough info for lineItem mapping for klarna ?

Reply:
Same idea as above. In the docs it says: `Set automatically once the ShippingMethod is set.`. Once you make a payment, I would expect ShippingMethod to be set.

Reply:
https://github.com/commercetools/commercetools-adyen-integration/pull/294/commits/e1fb9e8fec635a465b3b350798818a0c3279bee8